ascii2uni.1
.TH ascii2uni 1  "December, 2010"
.SH NAME
ascii2uni \- convert 7-bit ASCII representations to UTF-8 Unicode
.SH SYNOPSIS
.B ascii2uni [options] (<input file name>)
.SH DESCRIPTION
.I ascii2uni
converts various 7-bit ASCII representations to UTF-8.  It reads from the standard
input and writes to the standard output. The representations understood
are listed below under the command line options. If no format is specified, standard
hexadecimal format (e.g. 0x00e9) is assumed.
.PP

.SH "COMMAND LINE OPTIONS"
.sp 1
.B \-a <format>
Convert from the specified format. Formats may be specified by means of the following
arbitrary single character codes, by means of names such as "SGML_decimal", and by
examples of the desired format.
.IP
.B A
Convert hexadecimal numbers with prefix U in angle-brackets (<U00E9>).
.IP
.B B
Convert \\x-escaped hex (e.g. \\x00E9)
.IP
.B C
Convert \\x escaped hexadecimal numbers in braces (e.g. \\x{00E9}).
.IP
.B D
Convert decimal HTML numeric character references (e.g. &#0233;)
.IP
.B E
Convert hexadecimal with prefix U (U00E9).
.IP
.B F
Convert hexadecimal with prefix u (u00E9).
.IP
.B G
Convert hexadecimal in single quotes with prefix X (e.g. X'00E9').
.IP
.B H
Convert hexadecimal HTML numeric character references (e.g. &#x00E9;)
.IP
.B I
Convert hexadecimal UTF-8 with each byte's hex preceded by an =-sign (e.g. =C3=A9) . This is the 
Quoted Printable format defined by RFC 2045. 
.IP
.B J
Convert hexadecimal UTF-8 with each byte's hex preceded by a %-sign (e.g.  %C3%A9). This is the URIescape format defined by RFC 2396. 
.IP
.B K
Convert octal UTF-8 with each byte escaped by a backslash (e.g.  \\303\\251)
.IP
.B L
Convert \\U-escaped hex outside the BMP, \\u-escaped hex within the BMP (U+0000-U+FFFF).
.IP
.B M
Convert hexadecimal SGML numeric character references (e.g. \\#xE9;)
.IP
.B N
Convert decimal SGML numeric character references (e.g. \\#233;)
.IP
.B O
Convert octal escapes for the three low bytes in big-endian order(e.g. \\000\\000\\351))
.IP
.B P
Convert hexadecimal numbers with prefix U+ (e.g. U+00E9)
.IP
.B Q
Convert HTML character entities (e.g. &eacute;).
.IP
.B R
Convert raw hexadecimal numbers (e.g. 00E9)
.IP
.B S
Convert hexadecimal escapes for the three low bytes in big-endian order (e.g. \\x00\\x00\\xE9)
.IP
.B T
Convert decimal escapes for the three low bytes in big-endian order (e.g. \\d000\\d000\\d233)
.IP
.B U
Convert \\u-escaped hexadecimal numbers (e.g. \\u00E9).
.IP
.B V
Convert \\u-escaped decimal numbers (e.g. \\u00233).
.IP
.B X
Convert standard hexadecimal numbers (e.g. 0x00E9).
.IP
.B Y
Convert all three types of HTML escape: hexadecimal and decimal character references and character entities.
.IP
.B 0
Convert hexadecimal UTF-8 with each byte's hex enclosed within angle brackets (e.g.  <C3><A9>).
.IP
.B 1
Convert Common Lisp format hexadecimal numbers (e.g. #x00E9).
.IP
.B 2
Convert Perl format decimal numbers with prefix v (e.g. v233).
.IP
.B 3
Convert hexadecimal numbers with prefix $ (e.g. $00E9).
.IP
.B 4
Convert Postscript format hexadecimal numbers with prefix 16# (e.g. 16#00E9).
.IP
.B 5
Convert Common Lisp format hexadecimal numbers with prefix #16r (e.g. #16r00E9).
.IP
.B 6
Convert ADA format hexadecimal numbers with prefix 16# and suffix # (e.g. 16#00E9#).
.IP
.B 7
Convert Apache log format hexadecimal UTF-8 with each byte's hex preceded by a backslash-x (e.g.  \\xC3\\xA9). 
.IP
.B 8
Convert Microsoft OOXML format hexadecimal numbers with prefix _x and suffix _ (e.g. _x00E9_).
.IP
.B 9
Convert %\\u-escaped hexadecimal numbers (e.g. %\\u00E9).
.TP
.B \-h 
Help. Print the usage message and exit.
.TP
.B \-v 
Print program version information and exit.
.TP
.B \-m
Accept deprecated HTML entities lacking final semicolon, e.g. 
"&#x00E9" in place of "&#x00E9;".
.TP
.B \-p 
Pure. Assume that the input consists entirely of escapes except for arbitrary
(but non-null) amounts of separating whitespace.
.TP
.B \-q
Be quiet. Do not chat unnecessarily.
.sp 1
.TP
.B \-Z <format>
Convert input using the supplied format. The format
specified will be used as the format string in a call
to sscanf(3) with a single argument consisting of a pointer
to an unsigned long integer. For example, to obtain the same results
as with the \-U flag, the format would be: \\u%04X.
.PP
If the format is Quoted-Printable, although it is not strictly speaking
conversion of an ASCII escape to Unicode, in accordance with RFC 2045,
if an equal-sign occurs at the end of an input line, both the equal-sign
and the immediately following newline are skipped.
.PP
All options that accept hexadecimal input recognize both upper- and lower-case
hexadecimal digits.

.SH "EXIT STATUS"
.PP
The following values are returned on exit:

.IP "0 SUCCESS"
The input was successfully converted.

.IP "3 INFO"
The user requested information such as the version number or usage synopsis
and this has been provided.

.IP "5 BAD OPTION"
An incorrect option flag was given on the command line.

.IP "7 OUT OF MEMORY"
Additional memory was unsuccessfully requested.

.IP "8 BAD RECORD"
An ill-formed record was detected in the input.

.sp 1
.SH "SEE ALSO"
uni2ascii(1)
.sp 1
.SH AUTHOR
Bill Poser <billposer@alum.mit.edu>
.SH LICENSE
GNU General Public License


NEWS
2011-05-14 4.18
Fixed bug in uni2ascii in which in certain cases the subsitution
      count was too high, fixing Debian bug #626268.
Patched to handle situation in NetBSD which lacks getline.
Clarified semantics of pure option as converting characters
in ascii range other than space and newline.
Fixed bug in which this was not implemented correctly for UTF8 types.

2011-02-16 4.17
Added to uni2ascii the following conversions to nearest ascii equivalent:
       U+2022 bullet to 'o'
       U+00B7 middle dot to period
       U+0085 next line  to newline
       U+2028 line separator to newline

2010-12-12 4.16
Ascii2uni Q format now works again.
Added U+2033 to characters converted to nearest ascii equivalent by uni2ascii B option.

2010-08-29 4.15
Renamed endian.h u2a_endian.h to avoid conflict with external endian.h.
Removed getline() from ascii2uni.c since it is standard as of POSIX2008.

2009-08-04 4.14
Fixed a bug that interfered with the use of the Q format in uni2ascii.
Fixed bug in which ascification of U+2502 and U+2503 added double quote to output.
Fixed a bug in which -a S option generated a "Converted so many chars" line for each
character due to leaving in debugging code.

2009-04-22 4.13
Fixed Debian bug #511527 which caused the count of characters converted
to ASCII by uni2ascii to be excessive.</li>

2009-03-25 4.12
Both programs now allow the input file name to be specified on the command
line without redirection.

2008-10-02 4.11
Added support for %uXXXX format.
Added support for <XX><XX> format.

2008-08-30 version 4.10
Fixed bug that made Y argument to -a flag in ascii2uni a no-op.
Added documentation for Y argument to -a flag to man page for ascii2uni.
Corrected documentation for Q argument to -a flag in man page for ascii2uni.
The help flag for both programs now provides correct information about the
    Q argument to the -a flag.
Giving Y as the argument to the -a flag is now an error for uni2ascii.
The action summary is now more informative. The new version is incompatible
    with u2a, which I am no longer updating. The old action summary can be
    chosen at configuration time by using the option --disable-newsummary.
More informative version information is now provided.

2008-05-06 version 4.9
Fixed bug that produced bad output or segfault if a line
ended in the prefix to an escape, e.g. just an =-sign
in quoted-printable format.

In quoted-printable format if a line ends in an =-sign,
both the equal sign and the following newline are skipped
in accordance with RFC2045.

2008-05-04 version 4.8
Fixes serious bug in ascii2uni due to inclusion in release of test version.

2008-04-26 version 4.7
Added -P option to pass through untransformed Unicode.
Added -B option as shorthand for defxy.
Added characters to -e and -x options (see ChangeLog for details).
Since hardly anyone seems to use it, the GUI u2a is no longer being developed.

2008-03-25 version 4.6
Replaced fgets call with GNU getline so as to handle arbitrarily long input lines.
Added support for OOXML format (e.g. _x00E9_).
Fixed bug affecting BYTEQ, BYTED, and BYTEH formats.

2008-03-20 version 4.5
Error messages now include the line number.

Microsoft-style HTML entities lacking final semi-colon are now
passed on by default rather than converted. The new -m flag
causes them to be converted.

2008-01-14 version 4.4
Added -y option for single-character ascii approximations
as per patch provided by Jesse Peterson (jesse.peterson@exbiblio.com)
Changed license to GPL v.3


ascii2uni.py
#!/usr/bin/env python
# Time-stamp: <2008-03-10 12:05:26 poser>
#
# This is a filter that reads 7-bit ASCII containing \u-escaped Unicode
# and converts it to UTF-8 Unicode.
#
# Copyright (C) 2005 William J. Poser (billposer@alum.mit.edu)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Free Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
# or go to the web page:  http://www.gnu.org/licenses/gpl.txt.

import sys
import codecs

Version = '1.0'

argc = len(sys.argv)
if (argc > 1) and (sys.argv[1] == "-v"):
    sys.stdout.write("ascii2uni %s\n" % (Version))
    sys.stdout.write("Copyright 2005 William J. Poser\n")
    sys.stdout.write("Released under the terms of the GNU General Public License.\n")
    sys.exit(2)
else:
    sys.stdout.write("This program is a filter that converts from 7-bit ASCII containing\n\u-escaped Unicode (as used in Python and Tcl) and UTF-8 Unicode.\n")
    sys.stdout.write("So long as Python's Unicode implementation is restricted to 16 bits\nonly the Basic Multilingual Plane (Plane 0) is covered.\n")
    
outfile = codecs.getwriter('utf-8')(sys.stdout)
outfile.write(unicode(sys.stdin.read(),'unicode-escape'))
sys.exit(0)

